tiny_dnn/core/framework/tensor.h

```diff
@@ -0,0 +1,481 @@
+        return *this;
+    }
+
+#ifdef CNN_USE_DEFAULT_MOVE_CONSTRUCTORS
+    Tensor(Tensor&& other) = default;        // move ctor
+    Tensor &operator = (Tensor&&) = default; // move assign
+#else
+    Tensor(Tensor&& other) { // for VS2013 we need to manually implement these if we want to have move semantics
+        shape_ = std::move(other.shape_);
+        host_data_ = std::move(other.host_data_);
+        device_data_ = std::move(other.device_data_);
+        data_is_on_host_ = other.data_is_on_host_;
+        data_dirty_ = other.data_dirty_;
+        return *this;
+    }
+    Tensor &operator = (Tensor&& other) {
+        shape_ = std::move(other.shape_);
+        host_data_ = std::move(other.host_data_);
+        device_data_ = std::move(other.device_data_);
+        data_is_on_host_ = other.data_is_on_host_;
+        data_dirty_ = other.data_dirty_;
+}
+#endif
+
+    // Returns the tensor shape
+    const std::array<cnn_size_t, 4>& shape() const { return shape_; }
+    
+    // Returns the value of a specified index in the tensor.
+    // Checked version (throw exceptions for out-of-range error)
+    U& host_at(const cnn_size_t d0,
+          const cnn_size_t d1,
+          const cnn_size_t d2,
+          const cnn_size_t d3) {
+        return *host_ptr(d0, d1, d2, d3);
+    }
+
+    U host_at(const cnn_size_t d0,
+                const cnn_size_t d1,
+                const cnn_size_t d2,
+                const cnn_size_t d3) const {
+        return *host_ptr(d0, d1, d2, d3);
+    }
+
+    // Returns the pointer to a specified index in the tensor
+    // Checked version (throw exceptions for out-of-range error)
+    const U* host_ptr(const cnn_size_t d0,
+        const cnn_size_t d1,
+        const cnn_size_t d2,
+        const cnn_size_t d3) const {
+        if (d0 >= shape_[0] || d1 >= shape_[1] ||
+            d2 >= shape_[2] || d3 >= shape_[3]) {
+            throw nn_error("Access tensor out of range.");
+        }
+
+        return host_data() + (
+            shape_[1] * shape_[2] * shape_[3] * d0 +
+            shape_[1] * shape_[2] * d3 +
+            shape_[1] * d2 +
+            d1
+            );
+    }
+
+    U* host_ptr(const cnn_size_t d0,
+        const cnn_size_t d1,
+        const cnn_size_t d2,
+        const cnn_size_t d3) {
+        if (d0 >= shape_[0] || d1 >= shape_[1] ||
+            d2 >= shape_[2] || d3 >= shape_[3]) {
+            throw nn_error("Access tensor out of range.");
+        }
+
+        return mutable_host_data() + (
+            shape_[1] * shape_[2] * shape_[3] * d0 +
+            shape_[1] * shape_[2] * d3 +
+            shape_[1] * d2 +
+            d1
+            );
+    }
+
+    // zero-overhead version (same performance to raw pointer access.
+    // have an assertion for out-of-range error)
+    U& operator[] (const size_t index) {
+        return mutable_host_data()[index];
```

Comment:
There is a comment `zero-overhead`, but is it really so?
Inside `mutable_host_data`, I see `if` statement so I thought the performance of this method is slower than raw pointer access.